src/java/org/apache/cassandra/db/compaction/BackgroundCompactionRunner.java

```diff
@@ -0,0 +1,396 @@
+        // in different order and make each CFS have equal chance to be selected
+        ArrayList<ColumnFamilyStore> compactionRequestsList = new ArrayList<>(compactionRequests.keySet());
+        Collections.shuffle(compactionRequestsList, random);
+
+        for (ColumnFamilyStore cfs : compactionRequestsList)
+        {
+            List<SettableFuture<RequestResult>> promises = compactionRequests.remove(cfs);
+
+            logger.trace("Checking compaction request for {}.{} - there are {} awaiting promises",
+                         cfs.getKeyspaceName(),
+                         cfs.getTableName(),
+                         promises.size());
+
+            RequestResult completionStatus = canRunCompactionTasks(cfs, promises);
+            if (completionStatus != null)
+            {
+                promises.stream().filter(p -> !p.isDone()).forEach(p -> p.set(completionStatus));
+                continue;
+            }
+
+            logger.trace("Scheduling a background task check for {}.{} with {}",
+                         cfs.keyspace.getName(),
+                         cfs.name,
+                         cfs.getCompactionStrategy().getName());
+
+            Supplier<CompletableFuture<?>> compactionTask = Optional.ofNullable(lazyRunCompactionTasks(cfs))
+                                                                    .orElseGet(() -> lazyRunUpgradeTasks(cfs));
+
+            if (compactionTask != null)
+            {
+                ongoingCompactions.add(cfs);
+
+                compactionTask.get().handle((ignored, throwable) -> {
+                    ongoingCompactions.remove(cfs);
+
+                    if (throwable != null)
+                    {
+                        logger.warn(String.format("Aborting compaction of %s.%s due to error", cfs.getKeyspaceName(), cfs.getTableName()), throwable);
+                        handleCompactionError(throwable, cfs);
+
+                        promises.forEach(p -> p.setException(throwable));
+                    }
+                    else
+                    {
+                        logger.trace("Finished compaction for {}.{}", cfs.getKeyspaceName(), cfs.getTableName());
+
+                        // Since we have run at least one task and thus the set of sstables has changed,
+                        // check for new compaction possibilities
+                        requestCompaction(cfs);
+
+                        promises.forEach(p -> p.set(RequestResult.COMPLETED));
+                    }
+
+                    return null;
+                });
+            }
+            else
+            {
+                promises.forEach(p -> p.set(RequestResult.COMPLETED));
+            }
+        }
+    }
+
+    private boolean maybeScheduleNextCheck()
+    {
+        if (checkExecutor.getQueue().isEmpty())
+        {
+            try
+            {
+                checkExecutor.execute(this);
```

Comment:
The previous version also had a periodic catch-all request for every cfs. Do we still have that?

Reply:
Without the changes suggested in https://github.com/datastax/cassandra/pull/230#discussion_r683515129 it is not needed because after requesting a background compaction, we are guaranteed that at least background check is run after the request is made and the request is consumed with that run (and all promises are completed)